### PR TITLE
RAINCATCH-956

### PR DESCRIPTION
**Pull request by austincunningham**
## Motivation
Logger module interface for the Raincatcher 

## Description
A few sentences describing the overall goals of the pull request's commits.

## Progress
- [x] Logger interface
- [x] Unit tests
- [x] Integration
- [x] Documentation

### CODE REVIEW

File: cloud/datasync/src/SyncServer.ts [hunk 1]
```diff
@@ -4,6 +4,16 @@ import {setupGlobalHandlers, setupHandlers} from './crud-handlers/HandlerMapper'
 import SyncDataSetOptions from './options/SyncDatasetOptions';
 import SyncOptions from './options/SyncGlobalOptions';
 import SyncApi from './SyncApi';
+import {BunyanLogger,LOG_LEVEL} from '../../logger-cloud/src/index';
+//import {BunyanLogger,LOG_LEVEL} from '@raincatcher/logger-cloud'; this don't work
+
+// constructor for loggers
+const logD = new BunyanLogger(LOG_LEVEL.DEBUG);
```
Comment (wtrocki) [inline cloud/datasync/src/SyncServer.ts]: Replace this with single instance.

File: cloud/logger-cloud/src/Logger.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+export interface Logger {
+    //log(message:string, object:any): void;
+    debug(message:string, object:any): void;
+    error(message:string, object:any): void;
+    info(message:string, object:any): void;
+    warn(message:string, object:any): void;
+}
+
+export enum LOG_LEVEL {
```
Comment (wtrocki) [inline cloud/logger-cloud/src/Logger.ts]: Not sure if we need this.

File: cloud/datasync/src/SyncServer.ts (hunk 1 above)
Comment (witmicko) [reply cloud/datasync/src/SyncServer.ts]: +1, check this https://github.com/witmicko/raincatcher-core/blob/7f4ce88a5ce6313c1cfb4ce283dbc2ed1e60b3fd/cloud/logger-debug/src/index.ts

try to be more verbose in var naming

File: cloud/datasync/src/SyncServer.ts [hunk 3]
```diff
@@ -17,10 +27,12 @@ export const SyncServer: SyncApi = {
    */
     connect(options: SyncOptions, callback: (err: any) => void) {
     if (options.globalSyncOptions) {
+      logI.info('SyncOptions set :', options.globalSyncOptions);
       sync.setConfig(options.globalSyncOptions);
     }
     const sdo = options.datasetConfiguration;
```
Comment (witmicko) [inline cloud/datasync/src/SyncServer.ts]: naming, not sure what `sdo` is

File: cloud/datasync/src/SyncServer.ts [hunk 4]
```diff
@@ -4,6 +4,16 @@ import {setupGlobalHandlers, setupHandlers} from './crud-handlers/HandlerMapper'
 import SyncDataSetOptions from './options/SyncDatasetOptions';
 import SyncOptions from './options/SyncGlobalOptions';
 import SyncApi from './SyncApi';
+import {BunyanLogger,LOG_LEVEL} from '../../logger-cloud/src/index';
+//import {BunyanLogger,LOG_LEVEL} from '@raincatcher/logger-cloud'; this don't work
```
Comment (paolobueno) [inline cloud/datasync/src/SyncServer.ts]: run bootstrap and it should work, we can't keep references to outside `datasync`

File: cloud/datasync/src/SyncServer.ts (hunk 1 above)
Comment (paolobueno) [reply cloud/datasync/src/SyncServer.ts]: +1, multiple instances should be for prefixing messages and so on, see https://github.com/trentm/node-bunyan#logchild and the debug function: https://github.com/visionmedia/debug

File: cloud/logger-cloud/src/Logger.ts (hunk 2 above)
Comment (witmicko) [reply cloud/logger-cloud/src/Logger.ts]: I think this is valid use case for this https://github.com/witmicko/raincatcher-core/blob/7f4ce88a5ce6313c1cfb4ce283dbc2ed1e60b3fd/cloud/logger-debug/src/index.ts#L46

File: cloud/logger-cloud/src/Logger.ts [hunk 5]
```diff
@@ -0,0 +1,11 @@
+
+
+export interface Logger {
+    //log(message:string, object:any): void;
+    debug(message:string, object:any): void;
```
Comment (paolobueno) [inline cloud/logger-cloud/src/Logger.ts]: I think we can remove the `: void` return values here, might make it more convenient to implement it with a weird library that for some reason returns something.

File: cloud/logger-cloud/test/BunyanLogger.ts [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+
+const log = new BunyanLogger({name:"test"});
+
+
+describe('Expected results', () => {
+ it('should return the expected responses', () => {
+   log.debug('debug logger message', loggerObject);
+   log.error('error logger message', loggerObject);
+   log.info('info logger message', loggerObject);
+   log.warn('warn logger message', loggerObject);
```
Comment (paolobueno) [inline cloud/logger-cloud/test/BunyanLogger.ts]: In order for us to be able to test anything we probably need to have the interface receive an arbitrary stream (defaulting to stdout in implementations) that we could then mock and test, but that could be overkill for now

Comment (wtrocki) [reply cloud/logger-cloud/test/BunyanLogger.ts]: That's good usage for mocks and spies.

Comment (paolobueno) [reply cloud/logger-cloud/test/BunyanLogger.ts]: yeah it's definitely an interesting exercise if @austincunningham wants to go for it :D

File: cloud/logger-cloud/src/Logger.ts [hunk 7]
```diff
@@ -0,0 +1,11 @@
...
+    error(message:string, object:any): void;
+    info(message:string, object:any): void;
+    warn(message:string, object:any): void;
```
Comment (wtrocki) [inline cloud/logger-cloud/src/Logger.ts]: maybe we can use same interface as bunyan here:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/bunyan/index.d.ts#L113

File: cloud/logger/README.md [hunk 8]
```diff
@@ -0,0 +1,40 @@
+# RainCatcher Logger Cloud
+
+RainCatcher wrapper for the Bunyan npm. This module creates an interface
```
Comment (wtrocki) [inline cloud/logger/README.md]: This module creates logging facade for node.js that is used by all Raincatcher modules. 
Users can use their own loggers by wrapping them into provided interface. 
By default module provides Bunyan logger implementation

File: cloud/logger/README.md [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+that wraps the Bunyan npm
+
+RainCatcher code is written in [TypeScript](http://typescriptlang.org), uses [Mocha](https://mochajs.org/) for unit tests and [Istanbul](https://istanbul.js.org/) for test coverage checks.
```
Comment (wtrocki) [inline cloud/logger/README.md]: Let's remove this. That's not module related description.

File: cloud/logger/integration/index.ts [hunk 10]
```diff
@@ -0,0 +1,32 @@
+// integration of logger
+// npm run integration
+
+// import
+import {BunyanLogger} from '../src/BunyanLogger';
+
+// constructor needs at least name
+const log = new BunyanLogger({name: 'index'});
```
Comment (wtrocki) [inline cloud/logger/integration/index.ts]: const log: Logger = new BunyanLogger(..

File: cloud/logger/README.md [hunk 11]
```diff
@@ -0,0 +1,46 @@
+# RainCatcher Logger Cloud
+
+This module creates logging facade for node.js that is used by all Raincatcher modules.
```
Comment (wtrocki) [inline cloud/logger/README.md]: node.js `based applications`

File: cloud/logger/README.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+Users can use their own loggers by wrapping them into provided interface.
```
Comment (wtrocki) [inline cloud/logger/README.md]: provided `Logger`

File: cloud/logger/README.md [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+By default module provides [Bunyan](https://www.npmjs.com/package/bunyan) logger implementation.
```
Comment (wtrocki) [inline cloud/logger/README.md]: Please refer to bunyan documentation for more information?

File: cloud/logger/integration/index.ts [hunk 14]
```diff
@@ -0,0 +1,14 @@
+// integration of logger
+// npm run integration
```
Comment (wtrocki) [inline cloud/logger/integration/index.ts]: Remove comments

File: cloud/logger/integration/index.ts [hunk 15]
```diff
@@ -0,0 +1,14 @@
...
+
+import {BunyanLogger, Logger} from '../src/index';
+
+// constructor needs at least name
```
Comment (wtrocki) [inline cloud/logger/integration/index.ts]: Remove this comment

File: cloud/logger/integration/index.ts [hunk 16]
```diff
@@ -0,0 +1,14 @@
...
+const log: Logger = new BunyanLogger({name: 'index'});
+// by default debug logging is turned off to enable debug logging change constructor to
```
Comment (wtrocki) [inline cloud/logger/integration/index.ts]: Best to say that constructor accept bunyan options. Please refer to `link`

File: cloud/logger/integration/index.ts [hunk 17]
```diff
@@ -0,0 +1,14 @@
...
+// const log = new BunyanLogger({name: 'index', level: 'debug'});
+
+log.debug('debug logger message\n', {logger: 'debug'});
```
Comment (wtrocki) [inline cloud/logger/integration/index.ts]: replace `logger` with something less confusing `code: 300` etc.

File: cloud/logger/README.md [hunk 18]
```diff
@@ -0,0 +1,21 @@
+# RainCatcher Logger Cloud
+
+This module creates logging facade for node.js based applications that is used by all Raincatcher modules.
+Users can use their own loggers by wrapping them into provided interface.
+By default module provides [Bunyan](https://www.npmjs.com/package/bunyan) logger implementation.
+
+## Quick start
+
+Import the your logger implementation and the Logger interface to be used.
```
Comment (wtrocki) [inline cloud/logger/README.md]: Remove this topic.

File: cloud/logger/test/BunyanLogger.ts [hunk 19]
```diff
@@ -0,0 +1,29 @@
+/// <reference types="mocha" />
+'use strict';
+import * as bunyan from 'bunyan';
+import { expect } from 'chai';
+import * as proxyquire from 'proxyquire';
+
+// const BunyanLogger = proxyquire.noCallThru().load('../src/BunyanLogger', {
```
Comment (JameelB) [inline cloud/logger/test/BunyanLogger.ts]: Remove if not needed

File: cloud/logger/test/BunyanLogger.ts [hunk 20]
```diff
@@ -0,0 +1,29 @@
...
+
+import {BunyanLogger} from '../src/BunyanLogger';
+const loggerObject: object = {test: 'test'};
+const emptyObject: object = {};
+
+const log = new BunyanLogger({name: 'test'});
+
+// call the different end points for logger
+describe('Expected results', () => {
+ it('should return the expected responses', () => {
```
Comment (JameelB) [inline cloud/logger/test/BunyanLogger.ts]: `'should log the expected messages'` instead?

File: cloud/logger/integration/index.ts (hunk 15 above)
Comment (austincunningham) [reply cloud/logger/integration/index.ts]: done

File: cloud/logger/integration/index.ts (hunk 14 above)
Comment (austincunningham) [reply cloud/logger/integration/index.ts]: done

File: cloud/logger/integration/index.ts (hunk 16 above)
Comment (austincunningham) [reply cloud/logger/integration/index.ts]: done

File: cloud/logger/integration/index.ts (hunk 17 above)
Comment (austincunningham) [reply cloud/logger/integration/index.ts]: done

File: cloud/logger/README.md (hunk 18 above)
Comment (austincunningham) [reply cloud/logger/README.md]: done

File: cloud/logger/test/BunyanLogger.ts (hunk 19 above)
Comment (austincunningham) [reply cloud/logger/test/BunyanLogger.ts]: done

File: cloud/logger/test/BunyanLogger.ts (hunk 20 above)
Comment (austincunningham) [reply cloud/logger/test/BunyanLogger.ts]: done

**Review by wtrocki — DISMISSED** [review]
Great Job!

File: cloud/logger/README.md [hunk 21]
```diff
@@ -0,0 +1,22 @@
+# RainCatcher Logger Cloud
+
+This module creates logging facade for node.js based applications that is used by all Raincatcher modules.
+Users can use their own loggers by wrapping them into provided interface.
```
Comment (JameelB) [inline cloud/logger/README.md:4]: ..into `the` provided interface

File: cloud/logger/README.md [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+By default module provides Bunyan logger implementation.
```
Comment (JameelB) [inline cloud/logger/README.md]: By default, `this` module provides a Bunyan...

File: cloud/logger/README.md [hunk 23]
```diff
@@ -0,0 +1,22 @@
...
+Please refer to bunyan documentation for more information?
```
Comment (JameelB) [inline cloud/logger/README.md]: Please refer to `the Bunyan` ... Remove `?` at the end of sentence

File: cloud/logger/README.md [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+
+## Quick start
+
+Import the your logger implementation and the Logger interface to be used.
```
Comment (JameelB) [inline cloud/logger/README.md]: remove `your`

File: cloud/logger/integration/index.ts [hunk 25]
```diff
@@ -0,0 +1,12 @@
+
+import {BunyanLogger, Logger} from '../src/index';
+
+// constructor accept bunyan options. Please refer to https://www.npmjs.com/package/bunyan
+const log: Logger = new BunyanLogger({name: 'index'});
+// by default debug logging is turned off to enable debug logging change constructor to
```
Comment (JameelB) [inline cloud/logger/integration/index.ts:6]: ...turned `off. In order to` enable debug logging, change...
Should maybe add this comment above the Logger declaration.

File: cloud/logger/README.md [hunk 26]
```diff
@@ -0,0 +1,22 @@
+# RainCatcher Logger Cloud
+
+This module creates logging facade for node.js based applications that is used by all Raincatcher modules.
+Users can use their own loggers by wrapping them into provided interface.
+By default module provides [Bunyan](https://www.npmjs.com/package/bunyan) logger implementation.
```
Comment (JameelB) [inline cloud/logger/README.md:5]: By default, `the` module...

File: cloud/logger/README.md [hunk 27]
```diff
@@ -0,0 +1,22 @@
...
+
+
+## Quick start
+
+Import the your logger implementation and the Logger interface to be used.
```
Comment (JameelB) [inline cloud/logger/README.md:10]: remove `your`

**Review by JameelB — APPROVED** [review]

File: cloud/logger/README.md (hunk 26 above)
Comment (austincunningham) [reply cloud/logger/README.md:5]: done

File: cloud/logger/README.md (hunk 27 above)
Comment (austincunningham) [reply cloud/logger/README.md:10]: done

File: cloud/logger/integration/index.ts (hunk 25 above)
Comment (austincunningham) [reply cloud/logger/integration/index.ts:6]: done

**Pull request merged**